src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx

```diff
@@ -164,46 +164,43 @@ namespace xclcpuemhal2 {
     }
 
   }
- 
-  static void sigHandler(int sn, siginfo_t *si, void *sc)
-  {
-    switch(sn) {
-      case SIGSEGV:
-        {
-          saveDeviceProcessOutputs();
-          kill(0,SIGSEGV);
-          exit(1);
-          break;
-        }
-      case SIGFPE :
-        {
-          saveDeviceProcessOutputs();
-          kill(0,SIGTERM);
-          exit(1);
-          break;
-        }
-      case SIGABRT:
-        {
-          saveDeviceProcessOutputs();
-          kill(0,SIGABRT);
-          exit(1);
-          break;
-        }
-    case SIGUSR1:
-        {
-	  // One of the spawned processes died for some reason,
-	  //  kill all of the others and exit the host code
-	  saveDeviceProcessOutputs() ;
-	  std::cerr << "Software emulation of compute unit(s) exited unexpectedly" 
-		    << std::endl ;
-	  kill(0, SIGTERM) ; 
-	  exit(1) ;
-	  break ;
-        }
-      default:
-        {
-          break;
-        }
+
+  static void sigHandler(int sn, siginfo_t *si, void *sc) {
+    switch (sn) {
+    case SIGSEGV: {
+      saveDeviceProcessOutputs();
+      kill(0, SIGSEGV);
+      exit(1);
+      break;
+    }
+    case SIGFPE: {
+      saveDeviceProcessOutputs();
+      kill(0, SIGTERM);
+      exit(1);
+      break;
+    }
+    case SIGABRT: {
+      saveDeviceProcessOutputs();
+      kill(0, SIGABRT);
+      exit(1);
+      break;
+    }
+    case SIGCHLD: // Prevent infinite loop when the emulator dies
+      if (si->si_code != CLD_KILLED && si->si_code != CLD_DUMPED)
+        break;
+    case SIGUSR1: {
+      // One of the spawned processes died for some reason,
+      //  kill all of the others and exit the host code
+      saveDeviceProcessOutputs();
+      std::cerr << "Software emulation of compute unit(s) exited unexpectedly"
+                << std::endl;
+      kill(0, SIGTERM);
+      exit(1);
+      break;
+    }
+    default: {
+      break;
+    }
```

Comment:
I see. There was a problem in the original code with a random mix of spaces and tabs...
Curious use of all these useless `{}` almost everywhere. Specially strange with the `default:` case.
@stsoe is this the local coding style?

Reply:
No.  The coding style is early exit and no braces for single statement.

Reply:
@Ralender .  Maybe remove all the unnecessary curly braces?

I don't quite understand the exit(1), are you sure the host application should be killed?  Isn't it enough to kill the emulation process?

Reply:
All this seems to have been here before but I agree that it does not prevent us to fix the ugliness.
Perhaps you could start by preventing any non commented code in your team... :-)

Reply:
Removed the Braces.

About the exit(1) optimally we would want to properly return an error to he host code when the emulator dies.
However in specific situation i experienced where the emulator died proper error propagation is quite hard because it happens deep in nested macros that are used i various contexts.
I still think that killing the Host is better then staying stuck in an infinite loop.